### PR TITLE
Feature/recommendation with calendar booking

### PR DESCRIPTION
**Pull request by Aakash06**
1. Link Recommendation with Calendar.
2. Add Recommendation notification email.

### CODE REVIEW

File: public/javascripts/analysispage.js [hunk 1]
```diff
@@ -34,7 +34,7 @@ function analysis(startDate, EndDate) {
     pieChart(startDate, EndDate);
     columnChart(startDate, EndDate);
     lineGraph(startDate, EndDate);
-    leaderBoard(startDate, EndDate);
+    //leaderBoard(startDate, EndDate);
```
Comment (Aakash06) [inline public/javascripts/analysispage.js]: Code commented as we are not showing LeaderBoard right now in Portal.

File: app/models/RecommendationsRepository.scala [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+import reactivemongo.play.json.BSONFormats.BSONDateTimeFormat
+
+case class RecommendationInfo(email: Option[String],
+                              name: String,
+                              topic: String,
+                              recommendation: String,
+                              submissionDate: BSONDateTime,
+                              updateDate: BSONDateTime,
+                              approved: Boolean = false,
+                              decline: Boolean = false,
```
Comment (mavericksid) [inline app/models/RecommendationsRepository.scala]: Change "decline" to "declined"

File: app/models/RecommendationsRepository.scala [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+                              pending: Boolean = true,
+                              done: Boolean = false,
```
Comment (mavericksid) [inline app/models/RecommendationsRepository.scala]: Make sure done field is updated automatically once an Admin schedules a session.

File: app/models/RecommendationsRepository.scala [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+                              _id: BSONObjectID = BSONObjectID.generate())
+
+object RecommendationsJsonFormats {
+
+  import play.api.libs.json.Json
+
+  implicit val recommendationsFormat = Json.format[RecommendationInfo]
+}
+
+class RecommendationsRepository @Inject()(reactiveMongoApi: ReactiveMongoApi, dateTimeUtitlity: DateTimeUtility) {
```
Comment (mavericksid) [inline app/models/RecommendationsRepository.scala]: Correct typo "dateTimeUtitlity"

File: app/models/RecommendationsRepository.scala [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+      .flatMap(jsonCollection =>
+        jsonCollection
+          .insert(recommendationInfo))
+
+  def approveRecommendation(id: String)(implicit ex: ExecutionContext): Future[WriteResult] = {
+    val selector = BSONDocument("_id" -> BSONDocument("$oid" -> id))
+    val modifier = BSONDocument(
+      "$set" -> BSONDocument(
+        "approved" -> true,
+        "decline" -> false,
```
Comment (mavericksid) [inline app/models/RecommendationsRepository.scala]: Double check on other boolean fields that are required to be updated here.

File: app/models/RecommendationsRepository.scala [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        jsonCollection.update(selector, modifier))
+  }
+
+  def pendingRecommendation(id: String)(implicit ex: ExecutionContext): Future[UpdateWriteResult] = {
+    val selector = BSONDocument("_id" -> BSONDocument("$oid" -> id))
+
+    val modifier = BSONDocument(
+      "$set" -> BSONDocument(
+        "pending" -> true,
+        "done" -> false,
```
Comment (mavericksid) [inline app/models/RecommendationsRepository.scala]: Double check if more boolean fields are to be updated.

File: app/models/RecommendationsRepository.scala [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+        jsonCollection.update(selector, modifier))
+  }
+
+  def doneRecommendation(id: String)(implicit ex: ExecutionContext): Future[UpdateWriteResult] = {
+    val selector = BSONDocument("_id" -> BSONDocument("$oid" -> id))
+
+    val modifier = BSONDocument(
+      "$set" -> BSONDocument(
+        "pending" -> false,
+        "done" -> true,
```
Comment (mavericksid) [inline app/models/RecommendationsRepository.scala]: Double check if more boolean fields are to be updated.

File: app/models/RecommendationsRepository.scala [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+      .flatMap(jsonCollection =>
+        jsonCollection.update(selector, modifier))
+  }
+
+  def bookRecommendation(id: String)(implicit ex: ExecutionContext): Future[UpdateWriteResult] = {
+    val selector = BSONDocument("_id" -> BSONDocument("$oid" -> id), "approved" -> true)
+
+    val modifier = BSONDocument(
+      "$set" -> BSONDocument(
+        "book" -> true,
```
Comment (mavericksid) [inline app/models/RecommendationsRepository.scala]: Double check if more boolean fields are to be updated.

File: app/models/RecommendationsRepository.scala [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+      .flatMap(jsonCollection =>
+        jsonCollection.update(selector, modifier))
+  }
+
+  def cancelBookedRecommendation(id: String)(implicit ex: ExecutionContext): Future[UpdateWriteResult] = {
+    val selector = BSONDocument("_id" -> BSONDocument("$oid" -> id), "approved" -> true)
+
+    val modifier = BSONDocument(
+      "$set" -> BSONDocument(
+        "book" -> false,
```
Comment (mavericksid) [inline app/models/RecommendationsRepository.scala]: Double check if more boolean fields are to be updated.

File: app/models/RecommendationResponseRepository.scala [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+
+object RecommendationResponseJsonFormats {
+
+  import play.api.libs.json.Json
+
+  implicit val recommendationResponseFormat = Json.format[RecommendationResponseRepositoryInfo]
+}
+
+class RecommendationResponseRepository @Inject()(reactiveMongoApi: ReactiveMongoApi) {
```
Comment (mavericksid) [inline app/models/RecommendationResponseRepository.scala]: Rename to RecommendationsResponseRepository

File: app/models/RecommendationResponseRepository.scala [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+import reactivemongo.play.json.BSONFormats.BSONDateTimeFormat
+
+case class RecommendationResponseRepositoryInfo(email: String,
+                                                recommendationId: String,
+                                                upVote: Boolean,
+                                                downVote: Boolean,
+                                                _id: BSONObjectID = BSONObjectID.generate())
+
+
+object RecommendationResponseJsonFormats {
```
Comment (mavericksid) [inline app/models/RecommendationResponseRepository.scala]: Rename to RecommendationsResponseJsonFormats

File: app/controllers/RecommendationController.scala [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+  def validateTopic: Option[String] =
+    if (topic.isEmpty) {
+      Some("Topic must not be empty")
+    } else if (topic.length > 140) {
+      Some("Topic must be of 140 characters or less")
+    } else {
+      None
+    }
+
+  def validateRecommendation: Option[String] =
```
Comment (mavericksid) [inline app/controllers/RecommendationController.scala]: Change this to validateDescription.

File: app/controllers/RecommendationController.scala [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+import reactivemongo.bson.BSONDateTime
+import utilities.DateTimeUtility
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+case class Recommendation(email: Option[String],
+                          name: String,
+                          topic: String,
+                          recommendation: String,
```
Comment (mavericksid) [inline app/controllers/RecommendationController.scala]: Rename to description.

File: app/models/RecommendationsRepository.scala [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+import scala.concurrent.{ExecutionContext, Future}
+
+// this is not an unused import contrary to what intellij suggests, do not optimize
+import reactivemongo.play.json.BSONFormats.BSONObjectIDFormat
+import reactivemongo.play.json.BSONFormats.BSONDateTimeFormat
+
+case class RecommendationInfo(email: Option[String],
+                              name: String,
+                              topic: String,
+                              recommendation: String,
```
Comment (mavericksid) [inline app/models/RecommendationsRepository.scala]: Rename to description.

File: app/views/header.scala.html [hunk 15]
```diff
@@ -85,6 +88,12 @@
                     <span>Book a Session</span></a>
             </li>
 
+            <li class="@activeLink(routes.RecommendationController.renderRecommendationPage().url) alignLinks" data-toggle="tooltip" data-placement="right" title="Recommendation">
+                <a href="@routes.RecommendationController.renderRecommendationPage">
+                    <i class="fa fa-inbox"></i>
+                    <span>Recommendation</span></a>
```
Comment (mavericksid) [inline app/views/header.scala.html]: Rename this to "Suggest a Topic"

File: app/views/recommendations/recommendation.scala.html [hunk 16]
```diff
@@ -0,0 +1,152 @@
+@(email: Option[String])(implicit request: Request[_], messages: Messages, flash: Flash)
+
+@main("Recommendations") {
```
Comment (mavericksid) [inline app/views/recommendations/recommendation.scala.html]: Rename this to "Suggest a Topic"

File: app/views/recommendations/recommendation.scala.html [hunk 17]
```diff
@@ -0,0 +1,152 @@
+@(email: Option[String])(implicit request: Request[_], messages: Messages, flash: Flash)
```
Comment (mavericksid) [inline app/views/recommendations/recommendation.scala.html]: Move all inline styling to respective CSS file.

File: app/views/recommendations/recommendation.scala.html [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+                <label class="checkbox-outer filter-recommendation">
+                    <input type='radio' name='user-recommend-filter' checked class='custom-checkbox' value='all'/>
+                    <span class="filter-label_text"></span>
+                    <p class="filter-checkbox-text">All</p>
+                </label>
+            </div>
+        </div>
+
+        <div class="col-md-12">
+            <div data-bind="foreach: recommendation">
```
Comment (mavericksid) [inline app/views/recommendations/recommendation.scala.html]: Rename to recommendations.

File: public/javascripts/recommendation.js [hunk 19]
```diff
@@ -0,0 +1,311 @@
...
+        '</div>' +
+        '</form>',
+        buttons: {
+            formSubmit: {
+                text: 'Add',
+                btnClass: 'btn-blue',
+                action: function () {
+                    var username = this.$content.find('#user-name').val();
+                    var recommendationTopic = this.$content.find('#recommend-topic').val();
+                    var recommendation = this.$content.find('#recommend-text').val();
```
Comment (mavericksid) [inline public/javascripts/recommendation.js]: Rename to description and change alert as well.

File: public/javascripts/recommendation.js [hunk 20]
```diff
@@ -0,0 +1,311 @@
...
+
+    self.requestByUser = function (id) {
+        if (typeof(Storage) !== "undefined") {
+            sessionStorage.setItem("recommendationId", id);
+        }
+        window.location = jsRoutes.controllers.CalendarController.renderCalendarPage(true).url
+    };
+
+
+    function FetchRecommendationList(pageNumber, filter, sortBy) {
```
Comment (mavericksid) [inline public/javascripts/recommendation.js]: Use cameCase.

File: public/javascripts/recommendation.js [hunk 21]
```diff
@@ -0,0 +1,311 @@
+$(function () {
+    ko.applyBindings(new Recommendation());
+});
+
+function Recommendation() {
+
+    var page = 1;
+    var CSRFToken = $("#csrfToken").val();
```
Comment (mavericksid) [inline public/javascripts/recommendation.js]: Cross check it CSRFToken is not used, remove it!

**Pull request merged**